### PR TITLE
[SHELL32] Use STDMETHOD macro and keyword override

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

For simplicity and short typing. Macro `STDMETHOD` is defined in `<basetyps.h>` as follows:

```c
# define STDMETHOD(m) virtual HRESULT STDMETHODCALLTYPE m
# define STDMETHOD_(t,m) virtual t STDMETHODCALLTYPE m
```
C++11 keyword `override` is used for checking whether the method is overrided.

JIRA issue: [CORE-19469](https://jira.reactos.org/browse/CORE-19469)

## Proposed changes

- Replace `virtual HRESULT STDMETHODCALLTYPE m` with `STDMETHOD(m)` (`m` is a method name).
- Replace `virtual t STDMETHODCALLTYPE m` with `STDMETHOD_(t, m)` (`t` is a type. `m` is a method name).
- Use `override` keyword as possible.
- `CDefView` should inherit `IShellView3` due to override `CreateViewWindow3` method (really?).
- Fix `CDefView::CreateViewWindow3` (parameter `prcView` is `const RECT *`, not `LPRECT`).

## TODO

- [x] Do build.

### CODE REVIEW

**Comment by whindsaks** [comment]
If you want to follow the Microsoft convention, it should be `STDMETHOD(Foo)(THIS)PURE;` where PURE is override or =0 or nothing depending on the compiler or C vs C++. THIS and THIS_ is only relevant for C code.

File: dll/win32/shell32/droptargets/CRecyclerDropTarget.cpp [hunk 1]
```diff
@@ -109,8 +109,8 @@ class CRecyclerDropTarget :
             cfShellIDList = RegisterClipboardFormatW(CFSTR_SHELLIDLIST);
         }
 
-        HRESULT WINAPI DragEnter(IDataObject *pDataObject,
-                                            DWORD dwKeyState, POINTL pt, DWORD *pdwEffect)
+        STDMETHOD(DragEnter)(IDataObject *pDataObject,
```
Comment (HBelusca) [inline dll/win32/shell32/droptargets/CRecyclerDropTarget.cpp:112]: Refering to the other PR where I mentioned an article,
https://microsoft.public.vc.atl.narkive.com/j1pgALCm/stdmethod-vs-stdmethodimp#post2
> [...] Thus STDMETHOD must be used for declaring new methods (e.g. in
the interface definition produced by MIDL compiler), while STDMETHODIMP
must be used when implementing the method in .cpp file. Either can be
used when declaring a method in your class derived from the interface
('virtual' keyword is optional there). [...]

Here we are in the case of using `STDMETHODIMP`.

File: dll/win32/shell32/folders/CDesktopFolder.h [hunk 2]
```diff
@@ -48,48 +48,46 @@ class CDesktopFolder :
...
 
         // IContextMenuCB
-        virtual HRESULT WINAPI CallBack(IShellFolder *psf, HWND hwndOwner, IDataObject *pdtobj, UINT uMsg, WPARAM wParam, LPARAM lParam);
+        STDMETHOD(CallBack)(IShellFolder *psf, HWND hwndOwner, IDataObject *pdtobj, UINT uMsg, WPARAM wParam, LPARAM lParam) override;
 
         /*** IItemNameLimits methods ***/
 
-        STDMETHODIMP
-        GetMaxLength(LPCWSTR pszName, int *piMaxNameLen)
+        STDMETHOD(GetMaxLength)(LPCWSTR pszName, int *piMaxNameLen) override
```
Comment (HBelusca) [inline dll/win32/shell32/folders/CDesktopFolder.h:85]: see my comment above about STDMETHODIMP

File: dll/win32/shell32/folders/CFSFolder.h [hunk 3]
```diff
@@ -38,55 +38,53 @@ class CFSFolder :
...
 
         // IShellFolderViewCB
-        virtual HRESULT WINAPI MessageSFVCB(UINT uMsg, WPARAM wParam, LPARAM lParam);
+        STDMETHOD(MessageSFVCB)(UINT uMsg, WPARAM wParam, LPARAM lParam) override;
 
         /*** IItemNameLimits methods ***/
 
-        STDMETHODIMP
-        GetMaxLength(LPCWSTR pszName, int *piMaxNameLen)
+        STDMETHOD(GetMaxLength)(LPCWSTR pszName, int *piMaxNameLen) override
```
Comment (HBelusca) [inline dll/win32/shell32/folders/CFSFolder.h:82]: ditto

File: dll/win32/shell32/folders/CMyDocsFolder.h [hunk 4]
```diff
@@ -42,45 +42,43 @@ class CMyDocsFolder :
...
 
         // IPersistFolder2
-        virtual HRESULT WINAPI GetCurFolder(PIDLIST_ABSOLUTE * pidl);
+        STDMETHOD(GetCurFolder)(PIDLIST_ABSOLUTE * pidl) override;
 
         /*** IItemNameLimits methods ***/
 
-        STDMETHODIMP
-        GetMaxLength(LPCWSTR pszName, int *piMaxNameLen)
+        STDMETHOD(GetMaxLength)(LPCWSTR pszName, int *piMaxNameLen) override
```
Comment (HBelusca) [inline dll/win32/shell32/folders/CMyDocsFolder.h:76]: ditto

File: dll/win32/shell32/shellmenu/CMenuSite.h [hunk 5]
```diff
@@ -59,49 +59,49 @@ class CMenuSite :
...
 
     // IServiceProvider
-    virtual HRESULT STDMETHODCALLTYPE QueryService(REFGUID guidService, REFIID riid, void **ppvObject);
+    STDMETHOD(QueryService)(REFGUID guidService, REFIID riid, void **ppvObject) override;
 
 
     // Using custom message map instead
     virtual BOOL ProcessWindowMessage(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam, LRESULT &lResult, DWORD mapId = 0);
 
     // UNIMPLEMENTED
```
Comment (HBelusca) [inline dll/win32/shell32/shellmenu/CMenuSite.h:98]: These look like the methods of the `IDeskBarClient` interface.

File: dll/win32/shell32/shellmenu/CMenuToolbars.h [hunk 6]
```diff
@@ -152,16 +152,16 @@ class CMenuStaticToolbar :
...
-    virtual HRESULT InternalGetTooltip(INT iItem, INT index, DWORD_PTR dwData, LPWSTR pszText, INT cchTextMax);
-    virtual HRESULT InternalExecuteItem(INT iItem, INT index, DWORD_PTR dwData);
-    virtual HRESULT InternalPopupItem(INT iItem, INT index, DWORD_PTR dwData, BOOL keyInitiated);
-    virtual HRESULT InternalHasSubMenu(INT iItem, INT index, DWORD_PTR dwData);
-    virtual HRESULT InternalContextMenu(INT iItem, INT index, DWORD_PTR dwData, POINT pt);
+    STDMETHOD(InternalGetTooltip)(INT iItem, INT index, DWORD_PTR dwData, LPWSTR pszText, INT cchTextMax) override;
+    STDMETHOD(InternalExecuteItem)(INT iItem, INT index, DWORD_PTR dwData) override;
+    STDMETHOD(InternalPopupItem)(INT iItem, INT index, DWORD_PTR dwData, BOOL keyInitiated) override;
+    STDMETHOD(InternalHasSubMenu)(INT iItem, INT index, DWORD_PTR dwData) override;
+    STDMETHOD(InternalContextMenu)(INT iItem, INT index, DWORD_PTR dwData, POINT pt) override;
```
Comment (HBelusca) [inline dll/win32/shell32/shellmenu/CMenuToolbars.h:164]: Not sure whether it would be better to reserve the "STDMETHOD" thing _**just**_ for the overriden interface functions.
Here these are private/protected helpers of the class, so I'm not fully convinced of the change.

File: dll/win32/shell32/shellmenu/CMenuToolbars.h [hunk 7]
```diff
@@ -179,14 +179,14 @@ class CMenuSFToolbar :
...
```
Comment (HBelusca) [inline dll/win32/shell32/shellmenu/CMenuToolbars.h:191]: Same comment

File: dll/win32/shell32/shellmenu/CMergedFolder.h [hunk 8]
```diff
@@ -21,15 +21,15 @@
...
 interface IAugmentedShellFolder : public IShellFolder
 {
-    virtual HRESULT STDMETHODCALLTYPE AddNameSpace(LPGUID, IShellFolder *, LPCITEMIDLIST, ULONG) = 0;
-    virtual HRESULT STDMETHODCALLTYPE GetNameSpaceID(LPCITEMIDLIST, LPGUID) = 0;
-    virtual HRESULT STDMETHODCALLTYPE QueryNameSpace(ULONG, LPGUID, IShellFolder **) = 0;
-    virtual HRESULT STDMETHODCALLTYPE EnumNameSpace(ULONG, PULONG) = 0;
+    STDMETHOD(AddNameSpace)(LPGUID, IShellFolder *, LPCITEMIDLIST, ULONG) = 0;
+    STDMETHOD(GetNameSpaceID)(LPCITEMIDLIST, LPGUID) = 0;
+    STDMETHOD(QueryNameSpace)(ULONG, LPGUID, IShellFolder **) = 0;
+    STDMETHOD(EnumNameSpace)(ULONG, PULONG) = 0;
```
Comment (HBelusca) [inline dll/win32/shell32/shellmenu/CMergedFolder.h:27]: Maybe the `= 0` changed to `PURE` (here and below)

File: dll/win32/shell32/shellmenu/CMergedFolder.h [hunk 9]
```diff
@@ -99,115 +99,113 @@ class CMergedFolder :
...
-    STDMETHODIMP
-    GetMaxLength(LPCWSTR pszName, int *piMaxNameLen)
+    STDMETHOD(GetMaxLength)(LPCWSTR pszName, int *piMaxNameLen) override
     {
         return E_NOTIMPL;
     }
 
-    STDMETHODIMP
-    GetValidCharacters(LPWSTR *ppwszValidChars, LPWSTR *ppwszInvalidChars)
+    STDMETHOD(GetValidCharacters)(LPWSTR *ppwszValidChars, LPWSTR *ppwszInvalidChars) override
```
Comment (HBelusca) [inline dll/win32/shell32/shellmenu/CMergedFolder.h:208]: See comment about STDMETHODIMP

File: dll/win32/shell32/CDefView.cpp [hunk 10]
```diff
@@ -2953,7 +2963,8 @@ HRESULT STDMETHODCALLTYPE CDefView::CreateViewWindow3(IShellBrowser *psb, IShell
         TRACE("-- CommDlgBrowser\n");
     }
 
-    Create(m_hWndParent, prcView, NULL, WS_CHILD | WS_CLIPSIBLINGS | WS_CLIPCHILDREN | WS_TABSTOP, 0, 0U);
+    RECT rcView = *prcView;
+    Create(m_hWndParent, rcView, NULL, WS_CHILD | WS_CLIPSIBLINGS | WS_CLIPCHILDREN | WS_TABSTOP, 0, 0U);
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/CDefView.cpp:2967]: image
Won't work.

File: dll/win32/shell32/shellmenu/CMergedFolder.h (hunk 8 above)
Comment (katahiromz) [reply dll/win32/shell32/shellmenu/CMergedFolder.h]: Fixed.

File: dll/win32/shell32/shellmenu/CMergedFolder.h (hunk 9 above)
Comment (katahiromz) [reply dll/win32/shell32/shellmenu/CMergedFolder.h]: Fixed.

File: dll/win32/shell32/folders/CDesktopFolder.h (hunk 2 above)
Comment (katahiromz) [reply dll/win32/shell32/folders/CDesktopFolder.h]: Fixed.

File: dll/win32/shell32/folders/CFSFolder.h (hunk 3 above)
Comment (katahiromz) [reply dll/win32/shell32/folders/CFSFolder.h]: Fixed.

File: dll/win32/shell32/shellmenu/CMenuSite.h (hunk 5 above)
Comment (katahiromz) [reply dll/win32/shell32/shellmenu/CMenuSite.h]: Fixed.

File: dll/win32/shell32/shellmenu/CMenuToolbars.h (hunk 7 above)
Comment (katahiromz) [reply dll/win32/shell32/shellmenu/CMenuToolbars.h]: Fixed.

File: dll/win32/shell32/shellmenu/CMenuToolbars.h (hunk 6 above)
Comment (katahiromz) [reply dll/win32/shell32/shellmenu/CMenuToolbars.h]: Fixed.

File: dll/win32/shell32/folders/CMyDocsFolder.h (hunk 4 above)
Comment (katahiromz) [reply dll/win32/shell32/folders/CMyDocsFolder.h]: Fixed.

File: dll/win32/shell32/droptargets/CRecyclerDropTarget.cpp (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shell32/droptargets/CRecyclerDropTarget.cpp]: Fixed.

**Comment by katahiromz** [comment]
Double checked. Let's go.

File: dll/win32/shell32/CDefView.cpp [hunk 11]
```diff
@@ -97,7 +97,7 @@ struct MenuCleanup
 class CDefView :
     public CWindowImpl<CDefView, CWindow, CControlWinTraits>,
     public CComObjectRootEx<CComMultiThreadModelNoCS>,
-    public IShellView2,
+    public IShellView3,
```
Comment (learn-more) [inline dll/win32/shell32/CDefView.cpp:100]: This change does not seem related to `STDMETHOD` changes...

Comment (katahiromz) [reply dll/win32/shell32/CDefView.cpp:100]: I had embarrassed with deciding whether `CDefView::CreateViewWindow3` has to override. If it was overrided, then the class has to inherit `IShellView3`.
